src/tesla_client.cpp

```diff
@@ -0,0 +1,745 @@
+	  bool response_charge_enable_request = jresponse["charge_enable_request"]; // true
+
+	  int response_charge_limit_soc_max = jresponse["charge_limit_soc_max"]; // 100
+	  int response_charge_limit_soc_min = jresponse["charge_limit_soc_min"]; // 50
+	  int response_charge_limit_soc_std = jresponse["charge_limit_soc_std"]; // 90
+	  float response_charge_miles_added_ideal = jresponse["charge_miles_added_ideal"]; // 205.5
+	  bool response_charge_port_door_open = jresponse["charge_port_door_open"]; // true
+	  const char* response_charge_port_latch = jresponse["charge_port_latch"]; // "Engaged"
+	  int response_charge_rate = jresponse["charge_rate"]; // 0
+	  bool response_charge_to_max_range = jresponse["charge_to_max_range"]; // false
+	  int response_charger_actual_current = jresponse["charger_actual_current"]; // 0
+	  int response_charger_pilot_current = jresponse["charger_pilot_current"]; // 48
+	  int response_charger_power = jresponse["charger_power"]; // 0
+
+	  const char* response_charging_state = jresponse["charging_state"]; // "Complete"
+	  const char* response_conn_charge_cable = jresponse["conn_charge_cable"]; // "IEC"
+	  float response_est_battery_range = jresponse["est_battery_range"]; // 187.27
+	  const char* response_fast_charger_brand = jresponse["fast_charger_brand"]; // ""
+	  bool response_fast_charger_present = jresponse["fast_charger_present"]; // false
+	  const char* response_fast_charger_type = jresponse["fast_charger_type"]; // ""
+	  float response_ideal_battery_range = jresponse["ideal_battery_range"]; // 230.21
+	  bool response_managed_charging_active = jresponse["managed_charging_active"]; // false
+	  bool response_managed_charging_user_canceled = jresponse["managed_charging_user_canceled"]; // false
+	  int response_max_range_charge_counter = jresponse["max_range_charge_counter"]; // 0
+	  int response_minutes_to_full_charge = jresponse["minutes_to_full_charge"]; // 0
+	  bool response_not_enough_power_to_heat = jresponse["not_enough_power_to_heat"]; // false
+	  bool response_scheduled_charging_pending = jresponse["scheduled_charging_pending"]; // false
+
+	  long response_timestamp = jresponse["timestamp"]; // 1583079530271
+	  bool response_trip_charging = jresponse["trip_charging"]; // false
+	  int response_usable_battery_level = jresponse["usable_battery_level"]; // 83
+#endif // notyet
+	}
+      }
+      _activeRequest = TAR_NONE;
+    });
+}
+#else // use mongoose
+void TeslaClient::loop()
+{
+  if ((_activeRequest != TAR_NONE) &&
+      ((millis()-_lastRequestStart) > TESLA_REQ_TIMEOUT)) {
+    _activeRequest = TAR_NONE;
+  }
+
+  if (!_isBusy()) {
+    if ((millis()-_lastRequestStart) > TESLA_REQ_INTERVAL) {
+      if (_accessToken.length() == 0) {
+	requestAccessToken();
+      }
+      else if (_vehicleCnt == 0) {
+	requestVehicles();
+      }
+      else {
+	requestChargeState();
+      }
+    } 
+  } // !busy
```

Comment:
Not sure this is the best way of doing this, should really send the next request as soon as the previous one is complete, how about only putting the timer around getting the charge state:
```suggestion
  if (!_isBusy()) {
      if (_accessToken.length() == 0) {
	requestAccessToken();
      }
      else if (_vehicleCnt == 0) {
	requestVehicles();
      }
      else if ((millis()-_lastRequestStart) > TESLA_REQ_INTERVAL)
       {
	requestChargeState();
      }
  } // !busy
```

Reply:
The problem with your suggested fix is that when the credentials are entered incorrectly, or the Tesla servers are just having issues, we're going to be hitting up their servers at a very high rate. Do you really want to do that? If so, I'll commit your suggestion

Reply:
True, should probably back off a little under error conditions, Ideally these should be feed back to the UI, but we are not very good at that anywhere :(